oshi-core/src/main/java/oshi/util/Memoizer.java

```diff
@@ -0,0 +1,167 @@
+        // Adapted from Guava's ExpiringMemoizingSupplier
+        return new Supplier<T>() {
+            final Supplier<T> delegate = original;
+            volatile T value;
+            volatile long expirationNanos;
+
+            @Override
+            public T get() {
+              long nanos = expirationNanos;
+              long now = System.nanoTime();
+              if (nanos == 0 || now - nanos >= 0) {
+                synchronized (this) {
+                  if (nanos == expirationNanos) { // recheck for lost race
+                    T t = delegate.get();
+                    value = t;
+                    nanos = now + ttlNanos;
+                    expirationNanos = (nanos == 0) ? 1 : nanos;
+                    return t;
+                  }
+                }
+              }
+              return value;
+            }
+        };
+    }
+
+    public static <T> Supplier<T> memoize2(Supplier<T> original) {
+        return memoize(original, Integer.MAX_VALUE);
+    }
+
+    /**
+     * Store a supplier in a delegate function to be computed only once.
+     * 
+     * @param original
+     *            The {@link java.util.function.Supplier} to memoize
+     * @return A memoized version of the supplier
+     */
+    public static <T> Supplier<T> memoize(Supplier<T> original) {
+        // Method copied from:
+        // https://stackoverflow.com/questions/35331327/
+        // does-java-8-have-cached-support-for-suppliers/35335467#35335467
+        return new Supplier<T>() {
+            Supplier<T> delegate = this::firstTime;
+            boolean initialized;
+            public T get() {
+                return delegate.get();
+            }
+
+            private synchronized T firstTime() {
+                if (!initialized) {
+                    T value = original.get();
+                    delegate = () -> value;
+                    initialized = true;
+                }
+                return delegate.get();
+            }
+        };
+    }
+
+    /*
+     * Methods suggested but don't work
+     */
+    public static <I, O> Function<I, O> memoize1(Function<I, O> f) {
+        return new Function<I, O>() {
+            private volatile O value;
+
+            public O apply(I i) {
+                O value = this.value;
+                if (value == null) {
+                    synchronized (this) {
+                        value = this.value;
+                        if (value == null) {
+                            this.value = value = f.apply(i);
+                            if (value == null) {
+                                throw new AssertionError("Internal error. Function must not return null");
+                            }
+                        }
+                    }
+                }
+                return value;
+            }
+        };
+    }
+
+    public static <T> Supplier<T> memoize1(Supplier<T> s) {
+        return () -> memoize1(i -> s.get()).apply(null);
```

Comment:
> I just tried your DCL version of the memoizer... it doesn't work (run the main method in the latest commit) ...

Ahah :) Yeah, it's not DCL, I was just too quick to convert `Supplier` into a `Function`. Obviously the correct code is:
```java
public static <T> Supplier<T> memoize1(Supplier<T> s) {
  Function<Void, T> f = memoize1(i -> s.get());
  return () -> f.apply(null);
}
```
Notice how this time a function is created only once, while in my first attempt I was creating a new function (by calling `memoize1(Function<I, O> f)`) each time `Supplier.get()` was called.

The manual way you are using to check wether an implementation of a memoizer is correct or not is not automatic and can only find obvious logical problems (as the one addressed above), but it can't detect memory synchronization issues.

It's very easy to modify [`MemoizedObjectConcurrencyTest.java`](https://gist.github.com/stIncMale/2fce5f39aa37ac4584e0635ca6f67fe5) so that it can be used for the methods in `Memoizer`. I did this and checked `memoize(Supplier<T> original, long ttlNanos)` implementation you adopted from Guava, and I also checked the `memoize1(Supplier<T> s)` with the fix specified above. Both passed the test, which is good news! :)

Reply:
Nice!   I just did a bunch of microbenchmarks on the Guava version vs. the one from the SO post.  The Guava version takes half the time as the SO version for the first read, and only slightly longer for later reads... so I'll ditch the current version and go with the Guava-based version ... and yes, use your test class eventually.  That's tonight's project.  I've got the first set of classes complete, so I'll probably merge this PR and start on another group.

Reply:
Once again, @stIncMale, I want to thank you for your patience as I learn this. I wish I'd have taken the time to figure it out before, it actually makes so many things easier that I struggled with for a while!

I attempted to implement your test class, and had to make a few changes:
- Removed `@Nullable` annotation
- Replaced `var` with explicit typing (testing on JDK8)
- In the calculation of `actualNumberOfRefreshes` the code was asking for `s.value` but since I'm testing a method rather than an object, my only way to get that was to call `s.get()` which pre-increments value before returning it, so I decremented for the test.  (Not sure if that's correct or thread-safe to do!)

When I run the class as a test by itself, it passes.  However, if I run tests on my entire project, that test consistently fails:

> java.lang.AssertionError: ttlNanos=100000, minExpectedNumberOfRefreshes=2.0, maxExpectedNumberOfRefreshes=110.0, actualNumberOfRefreshes=1

I suspect it's my decrementing hack at fault here, but I'm not sure, and not sure of the appropriate workaround to get the test working.

Related, this test clearly takes 60 seconds of CPU time, and I'd like to avoid running it with every build.  Is there a standard way of triggering the test only when required?

Reply:
> I suspect it's my decrementing hack at fault here

No, `actualNumberOfRefreshes = s.get() - 1` is fine because the thread doing this is the only thread accessing the `s` object (we know that all the other threads started in the executor `ex` are dormant because we waited for them to complete).

I think here is what happens:
- when you run this test from a full build, something terminates the test because all tests together (or the whole build) takes too much time (I did not look at how the build process is done in OSHI).
- when the test is terminated, executer's threads are interrupted
- interrupt is handled incorrectly in the test, it affects the `while` loop logic, when it should just throw an exception instead.

I was even able to reproduce the problem you are observing by simply adding a single line `Thread.currentThread().interrupt();` right before the `while` loop, which confirms my hypothesis.

The fix is simple, move the check of the interrupted flag inside the loop and throw an `InterruptedException`:
```
while (System.nanoTime() - beginNanos < iterationDurationNanos
    || (guaranteedLastIteration = !guaranteedLastIteration)) {
  if (Thread.currentThread().isInterrupted()) {
    throw new InterruptedException();
  }
  ...
}
```
I am confident this will solve the problem. It's also a correct way of handling interruptions in this test. I also fixed this test in the gist with the test that I linked earlier.

> Is there a standard way of triggering the test only when required?

I believe this is the standard way: https://github.com/junit-team/junit4/wiki/categories